internal/cri/server/sandbox_portforward_windows.go

```diff
@@ -17,61 +17,102 @@
 package server
 
 import (
-	"bytes"
 	"context"
 	"fmt"
 	"io"
+	"net"
+	"time"
 
-	"k8s.io/utils/exec"
+	"github.com/containerd/log"
 
-	sandboxstore "github.com/containerd/containerd/v2/internal/cri/store/sandbox"
-	cioutil "github.com/containerd/containerd/v2/pkg/ioutil"
+	netutils "k8s.io/utils/net"
 )
 
 func (c *criService) portForward(ctx context.Context, id string, port int32, stream io.ReadWriter) error {
-	stdout := cioutil.NewNopWriteCloser(stream)
-	stderrBuffer := new(bytes.Buffer)
-	stderr := cioutil.NewNopWriteCloser(stderrBuffer)
-	// localhost is resolved to 127.0.0.1 in ipv4, and ::1 in ipv6.
-	// Explicitly using ipv4 IP address in here to avoid flakiness.
-	cmd := []string{"wincat.exe", "127.0.0.1", fmt.Sprint(port)}
-	err := c.execInSandbox(ctx, id, cmd, stream, stdout, stderr)
+	sandbox, err := c.sandboxStore.Get(id)
 	if err != nil {
-		return fmt.Errorf("failed to execute port forward in sandbox: %s: %w", stderrBuffer.String(), err)
+		return fmt.Errorf("failed to find sandbox %q in store: %w", id, err)
 	}
-	return nil
-}
 
-func (c *criService) execInSandbox(ctx context.Context, sandboxID string, cmd []string, stdin io.Reader, stdout, stderr io.WriteCloser) error {
-	// Get sandbox from our sandbox store.
-	sb, err := c.sandboxStore.Get(sandboxID)
-	if err != nil {
-		return fmt.Errorf("failed to find sandbox %q in store: %w", sandboxID, err)
+	var podIP string
+	if !hostNetwork(sandbox.Config) {
+		// get ip address of the sandbox
+		podIP, _, err = c.getIPs(sandbox)
+		if err != nil {
+			return fmt.Errorf("failed to get sandbox ip: %w", err)
+		}
+	} else {
+		// port forwarding is not supported for host process containers
+		return fmt.Errorf("port forwarding not supported for host process containers")
 	}
 
-	// Check the sandbox state
-	state := sb.Status.Get().State
-	if state != sandboxstore.StateReady {
-		return fmt.Errorf("sandbox is in %s state", fmt.Sprint(state))
-	}
+	err = func() error {
+		var conn net.Conn
+		if netutils.IsIPv4String(podIP) {
```

Comment:
This probably wouldn't happen, and I would need to check. If the CNI returns an empty string for element 0 in the slice for eth0. This would fail and give a hard to troubleshoot error since it goes down the ipv6 path. I think this is more to highlight that I need to improve the logic in sandbox_run.go.